compiler/p/codegen/PPCBinaryEncoding.cpp

```diff
@@ -56,46 +56,374 @@ namespace TR { class Register; }
+   {
+   TR_ASSERT_FATAL_WITH_INSTRUCTION(instr, (val & 0x3fu) == val, "0x%x is out-of-range for DCM/DQM field", val);
+   *cursor |= val << 10;
+   }
+
+static void fillFieldSH5(TR::Instruction *instr, uint32_t *cursor, uint32_t val)
+   {
+   TR_ASSERT_FATAL_WITH_INSTRUCTION(instr, (val & 0x1fu) == val, "0x%x is out-of-range for SH(5) field", val);
+   *cursor |= val << 11;
+   }
+
+static void fillFieldSH6(TR::Instruction *instr, uint32_t *cursor, uint32_t val)
+   {
+   TR_ASSERT_FATAL_WITH_INSTRUCTION(instr, (val & 0x3fu) == val, "0x%x is out-of-range for SH(6) field", val);
+   *cursor |= (val & 0x1fu) << 11;
+   *cursor |= (val & 0x20u) >> 4;
+   }
+
+static void fillFieldSIM(TR::Instruction *instr, uint32_t *cursor, uint32_t val)
+   {
+   TR_ASSERT_FATAL_WITH_INSTRUCTION(instr, isValidInSignExtendedField(val, 0x1fu), "0x%x is out-of-range for SIM field", val);
+   *cursor |= (val & 0x1f) << 16;
+   }
+
+static void fillFieldSI(TR::Instruction *instr, uint32_t *cursor, uint32_t val)
+   {
+   TR_ASSERT_FATAL_WITH_INSTRUCTION(instr, isValidInSignExtendedField(val, 0xffffu), "0x%x is out-of-range for SI field", val);
+   *cursor |= val & 0xffff;
+   }
+
+static void fillFieldSI5(TR::Instruction *instr, uint32_t *cursor, uint32_t val)
+   {
+   TR_ASSERT_FATAL_WITH_INSTRUCTION(instr, isValidInSignExtendedField(val, 0x1fu), "0x%x is out-of-range for SI(5) field", val);
+   *cursor |= (val & 0x1f) << 11;
+   }
+
+static void fillFieldUIM(TR::Instruction *instr, uint32_t *cursor, int32_t numBits, uint32_t val)
+   {
+   uint32_t fieldMask = (1u << numBits) - 1;
+   TR_ASSERT_FATAL_WITH_INSTRUCTION(instr, (val & fieldMask) == val, "0x%x is out-of-range for UIM(%d) field", val, numBits);
+   *cursor |= val << 16;
+   }
+
+static void fillFieldUI(TR::Instruction *instr, uint32_t *cursor, uint32_t val)
+   {
+   // TODO: This is a hack until PIC sites are reworked. Currently, the PIC site handling code
+   // assumes that the entire address is in the immediate of the instruction, so we can't chop it to
+   // 16 bits like we should.
+   if (!instr->cg()->comp()->isPICSite(instr))
+      TR_ASSERT_FATAL_WITH_INSTRUCTION(instr, (val & 0xffffu) == val, "0x%x is out-of-range for UI field", val);
+   *cursor |= val & 0xffff;
+   }
+
+static void fillFieldM6(TR::Instruction *instr, uint32_t *cursor, int32_t val)
+   {
+   TR_ASSERT_FATAL_WITH_INSTRUCTION(instr, (val & 0x3f) == val, "0x%x is out-of-range for ME(6)/MB(6) field", val);
+   *cursor |= (val & 0x1f) << 6;
+   *cursor |= (val & 0x20);
+   }
+
+static void fillFieldMB5(TR::Instruction *instr, uint32_t *cursor, int32_t val)
+   {
+   TR_ASSERT_FATAL_WITH_INSTRUCTION(instr, (val & 0x1f) == val, "0x%x is out-of-range for MB(5) field", val);
+   *cursor |= val << 6;
+   }
+
+static void fillFieldME5(TR::Instruction *instr, uint32_t *cursor, int32_t val)
+   {
+   TR_ASSERT_FATAL_WITH_INSTRUCTION(instr, (val & 0x1f) == val, "0x%x is out-of-range for ME(5) field", val);
+   *cursor |= val << 1;
+   }
```

Comment:
IMO these names are fine as is and trying to describe what these fields do in the instruction would be counterproductive. The binary encoder does not and _should not_ care what the instruction will do with the value it fills in and these fields should be treated as black boxes aside from that. Theoretically, a future version of the ISA could add a new instruction using these fields for a different purpose than any current instruction uses them for and ideally the binary encoder wouldn't need to care (obviously, this requires further work beyond the scope of this PR, but this is what I'm working towards).

That being said, looking back at the code, the definitions of these fields are only really described in the comments in `PPCOpsDefines.hpp`. Maybe adding documentation comments to these functions to define the field they're filling in would help?

Reply:
I don't have a strong opinion with this. If you keep these as they are, i would suggest naming MDS_mb and me  (lower case) below as in ISA book, in order to distinguish MB and mb, ME and me.

Reply:
I'm fine with renaming the field names themselves to make this clearer. Unfortunately, using underscores or lowercase would break the naming conventions for functions here, so I'd like to avoid that if possible. I also see that the lowercase `mb`/`me` fields are shared between MD-form and MDS-form instructions, so linking them specifically to MDS-form seems a bit strange to me.

Would you be alright with using just `MB` and `ME` for the M-form instructions and then using `MDM` for the `mb`/`me` field of a MD-form or MDS-form instruction (these two fields act identically from a binary encoding perspective, so I haven't actually differentiated between the two), with a note in the documentation explaining their connections to the ISA names of these fields?

Reply:
These fields have been renamed in 82b0b55fd91528fe8388494bdf4b1ab302b9c962. Addition of documentation comments is in progress.

Reply:
I've added Doxygen to the `fillFill*` helpers in b2950499bc1d357e846081714eaeace3cf6aa79b. The comment for `fillFieldMDM` in that commit points out the correct ISA names.